storage/storage_helpers.cpp

```diff
@@ -28,28 +28,26 @@ bool IsEnoughSpaceForDownload(MwmSize mwmSize)
          Platform::TStorageStatus::STORAGE_OK;
 }
 
-bool IsEnoughSpaceForDownload(MwmSize mwmSizeDiff, MwmSize maxMwmSize)
-{
-  // Mwm size is less than |maxMwmSize|. In case of map update at first we download updated map
-  // and only after that we do delete the obsolete map. So in such a case we might need up to
-  // |maxMwmSize| of extra space.
-  return IsEnoughSpaceForDownload(mwmSizeDiff + maxMwmSize);
-}
-
 bool IsEnoughSpaceForDownload(CountryId const & countryId, Storage const & storage)
 {
   NodeAttrs nodeAttrs;
   storage.GetNodeAttrs(countryId, nodeAttrs);
+
   return IsEnoughSpaceForDownload(nodeAttrs.m_mwmSize);
 }
 
 bool IsEnoughSpaceForUpdate(CountryId const & countryId, Storage const & storage)
 {
   Storage::UpdateInfo updateInfo;
-  
   storage.GetUpdateInfo(countryId, updateInfo);
-  MwmSize spaceNeedForUpdate = updateInfo.m_sizeDifference > 0 ? updateInfo.m_sizeDifference : 0;
-  return IsEnoughSpaceForDownload(spaceNeedForUpdate, storage.GetMaxMwmSizeBytes());
+
+  // In concurrent downloading process we should accumulate space for:
+  // - download and save all MWMs or Diffs = m_totalDownloadSizeInBytes
+  // - max MWM file size to apply diff patch (patches are applying one-by-one) = m_maxFileSizeInBytes
+  // - final size difference between old and new MWMs = m_sizeDifference
+
+  MwmSize const diff = updateInfo.m_sizeDifference > 0 ? updateInfo.m_sizeDifference : 0;
+  return IsEnoughSpaceForDownload(diff + updateInfo.m_totalDownloadSizeInBytes + updateInfo.m_maxFileSizeInBytes);
```

Comment:
for non-patches scenario the `m_totalDownloadSizeInBytes` is a new (updated) size, so it includes `diff` already, doesn't it?

Reply:
Well no in general, because diff is size difference of final files on disk, but new_mwm_size - old_mwm_size != download_diff_size.
ApplyDiff(old_mwm : 100M, diff : 10M) can produce 200M new_mwm

Reply:
Well, let's keep on future patch scenario here.

Reply:
Another issue is that for e.g. updating 20 files 100Mb each (2Gb total, 150Mb / 3Gb after update) a user will have to have 3Gb of free space.
But actually with 3 max concurrent downloads it should be much less:
1Gb (size difference) + 3*150Mb (temp space for concurrent dls) = 1.45Gb

Reply:
Алгоритм должен быть умным для обновлений: для bulk загрузок, когда пользователь тыкнул сразу во много файлов, надо сортировать файлы по размеру, выбирать топ больших по количеству одновременных загрузок. А если недостаточно места для параллельных загрузок, качать в один поток, пока не будет достаточно.

Сложно.

Можно ли упростить? Может вообще не проверять по тычку свободное место? А сообщать об ошибке уже при попытке загрузить конкретный файл и недостатке места для него?

Reply:
Ну у нас сейчас двойственность на платформах ..
- Android качает файлы поочередно в несколько потоков
- iOS тянет файлы системным загрузчиком сразу пачкой

Для массивного обновления сейчас аппа по крайней мере скажет, что не сможет обновить _все_ и тогда юзер ручками станет обновлять по одному нужному ему файлу. Если мы уберем проверку, она станет качать, может обновить _ненужные_ карты, а потом сказать ой на важной ..

Reply:
А вообще да. Патч улучшил эстимейт для пары файлов, но ухудшил для большой пачки ..

Reply:
>     * Android качает файлы поочередно в несколько потоков

Я по-моему проверял, что несколько файлов параллельно может качаться.

>на андроиде тоже параллельно, если не индивидуально карты выбирались для апдейта, а одной кнопкой

Reply:
> Well, let's keep on future patch scenario here. In download-replace logic, we need only m_totalDownloadSizeInBytes space.

Предлагаю пока сделать для сценария download-replace, формулу для сценария диффов закомментить в туду - когда механизм диффов будет готов, то поменять будет легко.

А для диффов можно на самом деле так:
`max(m_sizeDifference, m_totalDownloadSizeInBytes) + m_maxFileSizeInBytes`
Потому что до применения диффа он может занимать место, зарезервированное под разницу размеров new-old. А после применения дифф удаляется всё равно.

Reply:
Для учёта N одновременных загрузок апдейтов для download-replace:
`m_sizeDifference + min(сумма_размеров_N_самых_больших_файлов, m_totalDownloadSizeInBytes - m_sizeDifference)`
Для простоты можно `m_sizeDifference + min(N * m_maxFileSizeInBytes, m_totalDownloadSizeInBytes - m_sizeDifference)`, это всё равно будет лучше, чем по максимуму резервировать.

А для сценария диффов - параллельные загрузки от применения диффов сейчас наверное вообще не зависят? Т.е. загрузки продолжаются на полной параллельности/скорости, несмотря на то, что процесс применения диффов сильно отстаёт?

Reply:
В идеале, все качается на максимальный канал параллельно, диффы применяются последовательно.
Формулы красивые, надо понять сколько равно N.